src/datasource.ts

```diff
@@ -1,95 +1,112 @@
-  HttpMethod,
-  DataQueryRequest,
   ResponseMetadata,
-  isError,
-  Tuple,
-  MetaResponses,
-  FailResponse,
+  Responses,
   SuccessResponse,
-  InputQueryTarget,
-  AnnotationResponse,
-  AnnotationQueryOptions,
-  FilterRequest,
-  AssetsFilterRequestParams,
-  EventsFilterRequestParams,
+  Tab,
+  Tuple,
+  VariableQueryData,
 } from './types';
-import {
-  formQueriesForTargets,
-  getTimeseries,
-  reduceTimeseries,
-  promiser,
-  getLimitsWarnings,
-  getCalculationWarnings,
-  datapointsPath,
-  stringifyError,
-  getLabelsForTarget,
-} from './cdfDatasource';
-import { Connector } from './connector';
-import { TimeRange } from '@grafana/ui';
-import { ParsedFilter, QueryCondition } from './parser/types';
-import { datapointsWarningEvent, failedResponseEvent, TIMESERIES_LIMIT_WARNING } from './constants';
+import { applyFilters, getRequestId, toGranularityWithLowerBound } from './utils';
 
-const { Asset, Custom, Timeseries } = Tab;
-const { POST } = HttpMethod;
+const appEventsLoader = SystemJS.load('app/core/app_events');
 
-export default class CogniteDatasource {
+export default class CogniteDatasource extends DataSourceApi<
+  CogniteQuery,
+  CogniteDataSourceOptions
+> {
   /**
    * Parameters that are needed by grafana
    */
-  id: number;
   url: string;
-  name: string;
 
   project: string;
   connector: Connector;
 
-  /** @ngInject */
-  constructor(
-    instanceSettings: CogniteDataSourceSettings,
-    backendSrv: BackendSrv,
-    private templateSrv: TemplateSrv
-  ) {
-    this.id = instanceSettings.id;
-    this.url = instanceSettings.url;
-    this.name = instanceSettings.name;
+  templateSrv: TemplateSrv;
+  backendSrv: BackendSrv;
+
+  constructor(instanceSettings: DataSourceInstanceSettings<CogniteDataSourceOptions>) {
+    super(instanceSettings);
+
     const { url, jsonData } = instanceSettings;
+    this.backendSrv = getBackendSrv();
+    this.templateSrv = getTemplateSrv();
+    this.url = url;
+    this.connector = new Connector(jsonData.cogniteProject, url, this.backendSrv);
     this.project = jsonData.cogniteProject;
-    this.connector = new Connector(jsonData.cogniteProject, url, backendSrv);
   }
 
   /**
    * used by panels to get timeseries data
    */
-  async query(options: QueryOptions): Promise<QueryResponse> {
+  async query(options: DataQueryRequest<CogniteQuery>): Promise<QueryResponse> {
     const queryTargets = filterEmptyQueryTargets(options.targets);
     let responseData = [];
-
     if (queryTargets.length) {
       try {
         const { failed, succeded } = await this.fetchTimeseriesForTargets(queryTargets, options);
         handleFailedTargets(failed);
         showWarnings(succeded);
         responseData = reduceTimeseries(succeded, getRange(options.range));
       } catch (error) {
-        console.error(error); // not sure it ever happens
+        /* eslint-disable-next-line no-console  */
+        console.error(error); // TODO: use app-events or something
```

Comment:
I was worred about this one, but it's fine to revert as the fix does not need to be part of this PR.